src/prof_recent.c

```diff
@@ -14,11 +14,13 @@
 #  define STATIC_INLINE_IF_NOT_TEST
 #endif
 
+typedef ql_head(prof_recent_t) prof_recent_head_t;
+
 ssize_t opt_prof_recent_alloc_max = PROF_RECENT_ALLOC_MAX_DEFAULT;
 malloc_mutex_t prof_recent_alloc_mtx; /* Protects the fields below */
 static atomic_zd_t prof_recent_alloc_max;
 static ssize_t prof_recent_alloc_count = 0;
-static prof_recent_t *prof_recent_alloc_dummy = NULL;
+static prof_recent_head_t prof_recent_alloc_head;
```

Comment:
This reminds me of something I had in mind long ago -

I used to have the idea of keeping two recent lists: one list of recent allocations, and another list of recent frees. The scenario where the recent frees could be useful would be object life time analysis, remote free analysis, etc. I deliberately removed the `alloc` in the list node name, because I wanted this `prof_recent_t` node to serve in both the recent allocation list and the recent free list.

However, upon a second thought, I don't think there's a good reason why they must be recent frees for us to do those analysis, if we compare with the alternative of keeping a list of "randomized recent allocations", where we always keep N records that are guaranteed to have been uniformly randomly chosen from all past allocations.

The disadvantage of the "randomized recent allocations" approach is that we have to throw away the allocations that are still living for those analysis. Are there any other disadvantages I miss?

The advantage is that it can be implemented as another mode on top of what we have, rather than requiring maintaining another brand new list. Another advantage is that the N randomly chosen allocations have the best "representativeness" in some sense; we cannot say anything about the last N frees in terms of their "representativeness".

If we agree that there's little value in the recent frees, I can do a more thorough renaming, to change all the `recent_alloc`s to just `recent`s.

Thoughts?

Reply:
I agree that maintaining a collection of "N randomly selected sampled allocations" makes more sense than the N most recently freed if we want whole-program lifetime statistics (and we should probably use reservoir sampling for that, so there's no need to reuse this list for it - we should preallocate an array of the appropriate size at boot).

I'm neutral-to-negative on whether that calls for a rename; `prof_recent_allocs` (or `prof_recent_alloc_list` or whatever) feels more descriptive to me than just `prof_recent` (which doesn't describe the contents at all). But that's not a strong belief; if you're very in favor of omitting `alloc` I'll still stamp.

Reply:
OK will change to `prof_recent_alloc_list`.

Agree that we should have multiple lists. We may always want the "most recent allocations" list for OOM investigation purpose.

Upon a second thought, I feel "randomized recent frees" might be better than "randomized recent allocations", for the following reasons:
- One of my concerns earlier was the complications of maintaining multiple lists, in particular the multiple pointer pairs I had to create, one for each list. (The pointer pairs mean a pointer from the object to the record and another one from the record back to the object; the pair is needed because any one of the two may die before the other and needs to notify the other.) If we only record the already released objects, then there's no need for the pointer pairs. It's more like the `prof_log` case (but with an upper limit and without the extra effort of maintaining the traces and threads).
- We don't need to throw away records for the analysis. This isn't a big deal but is just more sensible logically: e.g. for analyzing remote frees, why not just supply the frees.

Another interesting aspect that just comes up to me is around the "representativeness" I was claiming about. I think I was wrong. The records we collected are prof samples, so they're biased towards larger allocations. Therefore, our analysis on object life time or remote frees will not be equally representative across different allocation sizes. This can be problematic, because we care a lot about finding long-lived small allocations and remotely freed small allocations. The problem is orthogonal to whether we keep randomized N or last N. I think there might be a way to alter the reservoir sampling algorithm to over-sample the smaller allocations in a way that exactly neutralizes the bias.

Anyway, these are not relevant to this PR.